### ISSUE TITLE
Compatibility with new cuda and gpu

### ISSUE THREAD

**Issue by LiuStan99** [darrenjkt/MS3D#21]
Dear authors,

Thanks a lot for sharing this interesting project! But when I was trying to run it, there was a compatibility problem with my GPU.

Specifically, I have an NVIDIA GeForce RTX 4090 which corresponds to **sm_89**. But when I am using the docker image provided by you, it is complaining that it does not support **sm_89**, which I think is probably caused by the outdated Cuda version (10.2) and Pytorch version (1.8.1) in the image. 

However, when I tried to modify the Dockerfile by updating the Cuda to a newer version with the corresponding Pytorch version, it could not build correctly. I have tried multiple combinations with different Cuda versions and the installation commands were copied from the [official website](https://pytorch.org/get-started/locally/). But none of them worked. Here is the error reported:
```
=> ERROR [ 8/13] RUN pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118                                                                                                                                                           0.7s
------                                                                                                                                                                                                                                                                            
 > [ 8/13] RUN pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118:                                                                                                                                                                      
#0 0.375 Looking in indexes: https://download.pytorch.org/whl/cu118
#0 0.604 ERROR: Could not find a version that satisfies the requirement torch (from versions: none)
#0 0.604 ERROR: No matching distribution found for torch
------
Dockerfile:36
--------------------
  34 |     # Install torch and torchvision
  35 |     # See https://pytorch.org/ for other options if you use a different version of CUDA
  36 | >>> RUN pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118
  37 |     
  38 |     # Install python packages
--------------------
ERROR: failed to solve: process "/bin/sh -c pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118" did not complete successfully: exit code: 1
```
Here is the Dockerfile after modification:
```
FROM nvidia/cuda:11.8.0-devel-ubuntu18.04

RUN echo 'debconf debconf/frontend select Noninteractive' | debconf-set-selections

# Install basics
RUN apt-get update -y \
    && apt-get install build-essential \
    && apt-get install -y apt-utils git curl ca-certificates bzip2 tree htop wget \
    && apt-get install -y libglib2.0-0 libsm6 libxext6 libxrender-dev bmon iotop g++ python3.7 python3.7-dev python3.7-distutils

# Install cmake v3.13.2
RUN apt-get purge -y cmake && \
    mkdir /root/temp && \
    cd /root/temp && \
    wget https://github.com/Kitware/CMake/releases/download/v3.13.2/cmake-3.13.2.tar.gz && \
    tar -xzvf cmake-3.13.2.tar.gz && \
    cd cmake-3.13.2 && \
    bash ./bootstrap && \
    make && \
    make install && \
    cmake --version && \
    rm -rf /root/temp

# Install python
RUN ln -sv /usr/bin/python3.7 /usr/bin/python
RUN wget https://bootstrap.pypa.io/get-pip.py && \
	python get-pip.py && \
	rm get-pip.py

# Install python packages
RUN PIP_INSTALL="python -m pip --no-cache-dir install" && \
    $PIP_INSTALL numpy==1.19.3 llvmlite numba 

# Install torch and torchvision
# See https://pytorch.org/ for other options if you use a different version of CUDA
RUN pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118

# Install python packages
RUN PIP_INSTALL="python -m pip --no-cache-dir install" && \
    $PIP_INSTALL tensorboardX easydict pyyaml scikit-image tqdm SharedArray six

WORKDIR /root

# Install Boost geometry
RUN wget https://jaist.dl.sourceforge.net/project/boost/boost/1.68.0/boost_1_68_0.tar.gz && \
    tar xzvf boost_1_68_0.tar.gz && \
    cp -r ./boost_1_68_0/boost /usr/include && \
    rm -rf ./boost_1_68_0 && \
    rm -rf ./boost_1_68_0.tar.gz 

# A weired problem that hasn't been solved yet
RUN pip uninstall -y SharedArray && \
    pip install SharedArray

RUN pip install spconv-cu118
```
Could you provide me with some hints on how to solve this? By the way, even if I tried to build the image using the original Dockerfile as you provided, it still failed because it could not find the base image. Here is the error when I tried to build without modifying the Dockerfile:
```
=> ERROR [internal] load metadata for docker.io/nvidia/cuda:10.2-cudnn7-devel-ubuntu18.04                                                                                                                                                                                   0.6s
 => [auth] nvidia/cuda:pull token for registry-1.docker.io                                                                                                                                                                                                                   0.0s
------
 > [internal] load metadata for docker.io/nvidia/cuda:10.2-cudnn7-devel-ubuntu18.04:
------
Dockerfile:1
--------------------
   1 | >>> FROM nvidia/cuda:10.2-cudnn7-devel-ubuntu18.04
   2 |     
   3 |     RUN echo 'debconf debconf/frontend select Noninteractive' | debconf-set-selections
--------------------
ERROR: failed to solve: nvidia/cuda:10.2-cudnn7-devel-ubuntu18.04: docker.io/nvidia/cuda:10.2-cudnn7-devel-ubuntu18.04: not found
```

Thank you very much!

**Comment by darrenjkt** [comment]
The dockerfile in this repo is copied from OpenPCDet. They've released a dockerfile for cuda11.6 [here](https://github.com/open-mmlab/OpenPCDet/blob/master/docker/cu116.Dockerfile). I don't install much more to run MS3D. Hopefully you'll be able to run MS3D just fine after building it. 

Let me know if it ends up working or has further issues. I can update this repo with that docker file if it works.

**Comment by LiuStan99** [comment]
Dear Darren,

Thanks a lot for the help! Using the [new Dockerfile from OpenPCDet](https://github.com/open-mmlab/OpenPCDet/blob/master/docker/cu116.Dockerfile) works for now in the sense that I can run inference using the pretrained models, perform ensembling and temporal tracking inside the container built from it. But there was a 'Segmentation fault (core dumped)' error when I was trying to perform temporal refinement. I am not sure whether it is caused by this new environment. I also have not tested other functions (e.g. visualization) yet. I will update you once I finished testing it.

Thanks!

**Comment by darrenjkt** [comment]
How much CPU RAM do you have? The seg fault could be due to the ensemble ps dict growing too large and taking up all your RAM. For the temporal refinement, different parts of the refinement adds modified versions of the ps labels as separate dict keys, maybe this tipped your RAM over. 

The computer I used for experiments had 64GB of CPU RAM. How many detectors are you using and what size of dataset are you auto-labelling? You could try using less pre-trained detectors or auto-label parts of your dataset at a time and see if that works out.

**Comment by LiuStan99** [comment]
Dear Darren,

Thank you very much for the help and instruction! Sorry for the late response, I have been trying to fix the problem. 

It turns out the problem has nothing to do with the new environment or RAM. It is just because the pretrained detectors are not working well on my data, so there is no detection in 'initial_peudo_labels.pkl' after ensembling, which in turn triggered the error. If I run the pipeline with better detectors from OpenPCDet (e.g. CenterPoint and VoxelNet), then it works fine. 

Regarding the environment, besides using the new Dockerfile, I also changed 'torch-scatter' to a newer version to adapt the new cuda. Here is the command:
```
pip install torch-scatter -f https://data.pyg.org/whl/torch-1.13.1+cu116.html
```

Thanks again! Please let me know if there is anything you want to test with regarding the new environment, I can also try to help.

**Comment by darrenjkt** [comment]
What is your data like? 
- What lidar was it captured from (incl number of beams if it's a ring-based lidar)? 
- Is it mounted on a vehicle or infrastructure? 

The detectors used in this repo, VoxelRCNN and PVRCNN are newer and perform better than CenterPoint and VoxelNet. VoxelRCNN (center) is exactly the same as CenterPoint in the first stage but the authors propose a 2nd stage voxel-based refinement. Same with PVRCNN (center) where 2nd stage is point-based refinement. 

What was the config file you used in OpenPCDet that you used?

**Comment by LiuStan99** [comment]
Dear Darren,

My data is collected from an OUSTER lidar which has 128 beams vertically. The vertical field of view is 40 degrees. It is mounted on a vehicle. The frequency is 10 Hz.

However, I did try to manually downsample the point cloud by reducing the number of rings to 64, so that it is similar to the lidar used by Lyft. The models and configs I was using to generate initial detections were [GT-trained VoxelRCNN (Anchor) adapted to Lyft](https://github.com/darrenjkt/MS3D/blob/main/tools/cfgs/target_lyft/ms3d_waymo_voxel_rcnn_centerhead.yaml) and [pretrained PV-RCNN++ (Anchor) for Lyft](https://github.com/darrenjkt/MS3D/blob/main/tools/cfgs/lyft_models/uda_pv_rcnn_plusplus_resnet_anchorhead.yaml).

When running the nuScenes-pretrained [VoxelNet](https://github.com/open-mmlab/OpenPCDet/blob/master/tools/cfgs/nuscenes_models/cbgs_voxel0075_voxelnext_doubleflip.yaml) and [CenterPoint](https://github.com/open-mmlab/OpenPCDet/blob/master/tools/cfgs/nuscenes_models/cbgs_voxel0075_res3d_centerpoint.yaml) on my data, I also used the similar way to reduce the number of rings to 32 so that the the point cloud was similar to Velodyne HDL-32E in nuScenes.
 
I will also try the detectors trained on nuScenes in this repo and use the same technique to reduce the number of rings. Once I get the result, I will update you. Any insights from you would also be very helpful!

Once again, thanks a lot!

**Comment by darrenjkt** [comment]
MS3D definitely should work on Ouster 128-beam lidars. We've managed to auto-label our Ouster OS1 lidar (128-beam, 45° vertical FOV) using MS3D with pre-trained Waymo and nuScenes models. You do not need to reduce the number of rings. Are you setting the appropriate SHIFT_COOR (the height of your lidar above the ground)? This value is very important to getting good detections.

This is the result of training a 3D detector on our auto-labelled Ouster OS1 data: https://youtu.be/gjOgirSks8I?si=q0aEhlXLVDiso-md&t=1497

If you send me an email, with a screenshot of your waymo dataset license agreement, I can send you the pre-trained models for Waymo. Give these a shot. If you have more issues, please copy-paste the config file that you are using.

**Comment by LiuStan99** [comment]
Dear Darren,

Thanks a lot for the help! I did adjust the SHIFT_COOR. I did some more tests and found the pre-trained PV_RCNN on Lyft from the model zoo works quite well. The pre-trianed Voxel_RCNN produces much fewer detections, and the GT-trained VoxelRCNN (Anchor) on Lyft barely gives any detections. I am retraining the Voxel_RCNN using the ensembled results produced by the PV_RCNN now.

I don't think I can use the pre-trained detectors for Waymo, because I am doing this for a company. But thank you very much anyway!

**Comment by darrenjkt** [comment]
That's great to hear! For your information, Lyft provides labels for vehicles on the road and often do not label parked vehicles. Depending on your usage, this may result in lower recall which is means you have less candidates to filter from in the temporal refinement step. 

So even though you use a 128-beam lidar, I would also recommend adding some nuscenes pre-trained models in the ensemble to increase the recall. Nuscenes models do surprisingly well when there's a decent amount of points on the object despite the difference in lidar points density. 

Better yet, train some models on higher resolution lidar datasets like Argoverse, Pandaset or Zenseact to use as pre-trained models for auto-labelling.

**Comment by LiuStan99** [comment]
Dear Darren,

Thanks a lot for the valuable insights! I will incorporate them in my following work. 

I do have one extra question: I understand that this nice framework can adapt different lidars and I do not need to manually reduce the number of beams for Ouster when training. But do you recommend reducing the number of beams when producing the initial pseudo labels (e.g. reducing the number 32 when using models pre-trained on nuScenes), just as a naive way to improve the quality of the initial pseudo labels before any adaptation? Or is it unnecessary?

Many thanks for the answer in advance!

**Comment by darrenjkt** [comment]
There have been some works that say downsampling the beams works but I think it's hard to generalize to different scan patterns. For example with 64-beam lidars, there can be lots of different configurations e.g. beams centered on the horizon, or evenly spaced beams across the vertical FOV (VFOV). If you downsample a horizon-centered 64-beam lidar with 25° VFOV to 32-beams by skipping every 2nd beam, it won't be similar to nuscenes as the nuscenes 32-beam lidar is evenly spaced across a large 40° VFOV. 

If you want the best of both worlds, use the same e.g. nuscenes pre-trained detector and feed it both the down-sampled (to e.g. 32 beam) and non down-sampled (128 beam), then fuse both predictions in the ensemble with our KBF. I think that this is the better approach in order to get the highest recall for auto-labelling. With our bev visualization tools you can see which one gives the better result too if you're interested to find out.

**Comment by LiuStan99** [comment]
Dear Darren,

Thanks a lot for the advice!  I will try it.
